Wait for started threads in QueuedThreadPool

gregw opened the pull request:
Fix #13100 by waiting for newly started threads when offering jobs to the queue.

gregw committed:
Spinwait for started threads in QueuedThreadPool

Fix #13100 by starting new threads before offering jobs to the queue and if the queue is full, we spin a while until we see at least some threads have started.

Also ensure that rejected jobs are closed.

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -789,47 +802,68 @@ public void execute(Runnable job)
+            {
+                // Spin wait to allow threads to start
+                Thread.onSpinWait();
+                if (_counts.get() != counts)
```
I think this is a good basic approach, that if the queue is full we spin waiting for any started threads to have been started.

HOWEVER! This is the wrong test for that. Counts will only be updated if a Runner completes running a job or if more jobs are queued (and they may be spinning as well).

We don't really have anything we can spin on to ensure a thread is started.

The only other approach I can think of is to start a special Runner that has a reference to the job.  When that Runner takes a job from the queue it will immediately offer the job... but that may still race with other threads adding to the queue.  Perhaps it just keeps the job reference and tries to queue it after every job it takes until it can and then nulls it.

Let me try that.

gregw replied:
I've implemented this approach, which I think it better. so this comment will be outdated.

gregw committed:
Use a PendingRunner if the job queue is full.

Fix #13100 by using a new Runner that holds the pending job until it removes a job from the queue.

Also ensure that rejected jobs are closed.

gregw commented:
@sbordet @lorban I've taken a totally different approach.  If the queue is full, but we are starting a thread, then I start a PendingRunner, we takes on the responsibility of queuing the pendingJob once it has taken a job.

I have no idea how to test this????

gregw committed:
Use a PendingRunner if the job queue is full.

Fix #13100 by using a new Runner that holds the pending job until it removes a job from the queue.

Also ensure that rejected jobs are closed.

gregw committed:
Use a PendingRunner if the job queue is full.

Fix #13100 by using a new Runner that holds the pending job until it removes a job from the queue.

Also ensure that rejected jobs are closed.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java:
```diff
@@ -333,8 +333,7 @@ private void execute(Runnable task)
         {
-            if (task instanceof Closeable)
-                IO.close((Closeable)task);
+            // ignored.
```
I'm 50:50 on this one.   Perhaps it is best not to close a job on a REE and let the caller decide?
We close jobs when they cannot be executed after execute returns so REE is no longer an option

thoughts?

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/ExecutorThreadPool.java:
```diff
@@ -263,7 +264,15 @@ public int getIdleThreads()
+        }
+        catch (RejectedExecutionException e)
+        {
+            ThreadPool.reject(command, e);
```
540:50 on this.  See above.

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -788,48 +778,69 @@ public boolean isLowOnThreads()
-        if (!_jobs.offer(job))
+        // The job queue is full!
+
+        // if we are starting a thread, we can use a PendingRunner
```
If we will start a thread, there is no need to spin waiting, we can just create a PendingRunner, which will add it to the queue when it can.

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -788,48 +778,69 @@ public boolean isLowOnThreads()
-            LOG.debug("queue {} startThread={}", job, startThread);
+        // Otherwise, if we have pending threads started, let's wait for them to start
+        while (_pending.get() > 0)
+            Thread.onSpinWait();
```
This spin is in case a thread started from a previous execute has not yet started.  It is not a 100% fix, as there is still a race between the thread saying it has started and it polling the job queue.

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -1239,4 +1271,55 @@ private void doRunJob(Runnable job)
+            {
+                // we took a job from the queue, so offer our pending job
+                if (_jobs.offer(_pendingJob))
+                    _pendingJob = null;
```
This is fine if the offer succeeds... but sub optimal if it fails as the job taken might be a blocking job and block for a long time, thus indefinitely delaying the pending job.

Not sure what else can be done with this approach?

gregw commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/thread/QueuedThreadPoolTest.java:
```diff
@@ -1158,6 +1160,59 @@ public void testRealistic() throws Exception
+        qtp.execute(tasks[0]);
+        qtp.execute(tasks[1]);
+        qtp.execute(tasks[2]);
+        qtp.execute(tasks[3]);
```
This test captures the essence of the need for this PR.    If you configure a QTP with 2 threads and a bounded queue of 2, then you expect to be able to submit 4 jobs without a REE.   Currently this fails on executing the 4th job as the second thread has yet to start.     This PR "fixes" that, but only by shrinking this window of the race.  I think the race still exists.

gregw commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/thread/QueuedThreadPoolTest.java:
```diff
@@ -1158,6 +1160,59 @@ public void testRealistic() throws Exception
+        // Allow one more thread to run
+        qtp.setMaxThreads(3);
+        // Can immediately execute the 5th task
+        qtp.execute(tasks[4]);
```
This is kind of just a static way to trigger the dynamic issue discussed above.

gregw commented:
@sbordet @lorban @lachlan-roberts 

I'm not 100% convinced by this approach to fix the problem, but I am 100% convinced it is a problem we need to fix.
If you configure a QTP of N threads with a bounded queue of M tasks, then you should be able to execute N+M jobs without a RejectedExecutionException.
Currently the delay between starting a thread and it polling the job queue means that we fail to add to the bound queue because the started thread has not yet taken from the queue.

This PR attempts to fix with 2 strategies: a) spin wait for pending threads to start; b) if starting a threads with a full queue, give it the job to append to the queue after it has taken a job from the queue.

The problem with this approach is that both strategies are still racy.  For a), there is still a window between a thread saying it has started and polling the job queue.  For b) the PendingRunner can race with another thread adding a job to the queue and thus indefinitely delay the pending job.

Perhaps rather than these racy strategies we should just have a pending job queue as well as the real job queue.  We can add some jobs to that queue if we have pending threads being started.  Then all runners will attempt to move jobs from the pending queue to the end of the real queue?

thoughts?

sbordet commented:
@gregw I like the pending queue approach better, although it has a fixed cost that we will always need to pay (check the pending queue and possibly enqueue).

Note that we will always need to maintain this queue: even when we create all `maxThreads` threads, they can be evicted and the problem will present again: the job queue maxes out, but we have to create more threads to replace the evicted ones.

sbordet reviewed (changes requested):
I like better a pending queue approach.

gregw commented:
@sbordet I think I've found a better way. We atomically know the number of threads and the queue length. If they are at max capacity we reject even before trying the queue. If they are not max, then we spin until we can add as we know there is capacity (modulo checking for other state changes).

I think this approach will be fully atomic and cost little. Stand by....

gregw committed:
Wait for starting thread for bound queue in QTP.

Always allow full capacity to be reached by waiting for a started thread in QTP.

gregw committed:
Wait for starting thread for bound queue in QTP.

Initial BAQ put implementation (nees tests)

gregw commented:
@sbordet I've implemented my new approach and it works either will put (if supported) else it warns and falls back to spinning.
I've done a quick implementation of put on BAQ, but it needs careful review and tests.

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/13100/qtpRejectedJobs

gregw committed:
Wait for starting thread for bound queue in QTP.

Tested BAQ

gregw committed:
Wait for starting thread for bound queue in QTP.

Tested BAQ

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -146,6 +148,7 @@ public void clear()
             {
                 _headLock.unlock();
             }
+            _notFull.signalAll();
```
Just `signal()` is enough.

gregw replied:
I don't think so. See other reply.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -770,54 +828,59 @@ public int getMaxCapacity()
     }
 
-    private boolean grow()
+    private void growTailHeadLocked()
```
We typically prefix "locked", so rename to `lockedHeadGrowTail()`?
But how can we just lock "head", this should lock both tail and head?
It does indeed, so this is just `lockedGrow()`.

gregw replied:
we we sometimes just have the head lock or the tail lock.  This requires both, so I thought that should be clear in the name???

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -185,14 +188,15 @@ public E poll()
                 _indexes[HEAD_OFFSET] = (head + 1) % _elements.length;
-                if (_size.decrementAndGet() > 0)
-                    _notEmpty.signal();
+                wasFull = signalSizeHeadLocked();
```
I think this method hides the logic too much.
I tried to remove it, and inlining its implementation makes the core clearer, and the inline is typically only 2-3 lines, so I prefer the little duplication over the need to know what this method is doing.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -201,6 +205,8 @@ public E poll()
     {
         long nanos = unit.toNanos(time);
+        E e;
+        boolean wasFull = false;
```
Unnecessary initialization.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -220,19 +226,21 @@ public E poll(long time, TimeUnit unit) throws InterruptedException
+        if (wasFull)
+        {
+            signal(_tailLock, _notFull);
+        }
```
Drop braces.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -389,13 +401,14 @@ public boolean offer(E e)
                 {
-                    if (!grow())
-                        return false;
+                    if (_size.get() >= _maxCapacity && _size.get() != _elements.length)
```
Here we are fully locked, no need to access `_size` twice.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -416,15 +429,7 @@ public boolean offer(E e)
-            {
-                _headLock.unlock();
-            }
+            signal(_headLock, _notEmpty);
```
Drop braces.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -470,8 +527,9 @@ public void add(int index, E e)
                 {
                     if (_indexes[TAIL_OFFSET] == _indexes[HEAD_OFFSET])
-                        if (!grow())
+                        if (_growCapacity <= 0 || _size.get() == _maxCapacity)
```
Brace nested `if` statements for clarity.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -514,25 +572,25 @@ public void add(int index, E e)
-        // The mechanism to await and signal when the queue is full is not implemented
-        throw new UnsupportedOperationException();
+        if (!offer(e, Integer.MAX_VALUE, TimeUnit.SECONDS))
+            throw new RejectedExecutionException();
```
I'd rather have a little duplication and use `_notFull.await()`, rather than throwing `RejectedExecutionException` (see JDK's `ArrayBlockingQueue`).

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -514,25 +572,25 @@ public void add(int index, E e)
     public E take() throws InterruptedException
     {
+        E e = null;
+        boolean wasFull = false;
```
Unnecessary initializer.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -633,6 +690,7 @@ public int drainTo(Collection<? super E> c, int maxElements)
             {
                 _headLock.unlock();
             }
+            _notFull.signalAll();
```
`signal()` is enough.

gregw replied:
Is it?  If we have multiple threads waiting on notFull, then signal will only start one and the rest may wait forever unless the Q fills up again.  
We could signal notFull on every take, but then that is more work and we'd have to do N signals here.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -764,7 +762,10 @@ public int getMaxAvailableThreads()
     public double getUtilizationRate()
     {
-        return (double)getUtilizedThreads() / (getMaxThreads() - getLeasedThreads());
+        int available = getMaxThreads() - getLeasedThreads();
+        if (available <= 0)
+            return available == 0 ? 1.0D : 0.0D;
+        return (double)getUtilizedThreads() / available;
```
Unnecessary? The original code was fine.

gregw replied:
There is a chance that leased threads == max threads (not a great config but possible).  This change avoids a divide by zero.

sbordet replied:
I don't think it is possible that leasedThreads==maxThreads, as there would be no "worker" threads, so it would be an illegal configuration, see `ThreadPoolBudget.check()`.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -804,32 +803,78 @@ public void execute(Runnable job)
+            }
+            catch (UnsupportedOperationException uoe)
+            {
+                LOG.warn("Detected bounded thread pool queue {} without put support", _jobs.getClass());
```
```suggestion
                LOG.warn("Detected bounded thread pool queue {} without put() support", _jobs.getClass());
```

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -804,32 +803,78 @@ public void execute(Runnable job)
+            {
+                addCounts(0, 1);
+                Thread.currentThread().interrupt();
+                throw new RejectedExecutionException(e);
```
Is this enough, or we should call `ThreadPool.reject()` to close the job?
Also in other places.

gregw replied:
That was my first change in this PR, to always reject (close) when we throw REE.  But we do have places in the code that already catch and do that.    So I ended up thinking that we should not tell the app the same thing twice - i.e. we should not both close and throw.  We should either throw in execute, or if we return from execute, then we close if we cannot do it.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -804,32 +803,78 @@ public void execute(Runnable job)
 
-        if (LOG.isDebugEnabled())
-            LOG.debug("queue {} startThread={}", job, startThread);
+        // Put was not supported, so we will just spin instead
```
I don't think this is a good idea.
We have already been reported in the past about excessive CPU consumption due to spin.
We tried `put()`, now we reject, no spinning.

gregw replied:
Hmmm maybe. We do warn and its not a bad spin and I do throttle.     So I'll leave it in for now...   Perhaps we need to see how many Q implementations do not have put?

sbordet replied:
I'd really take it out.
If someone complains, then we can add it back, but it's the rare case of a rare case that will bite us back.
I'd prefer fail fast and reject.
`parkNanos()` will also spin, so now we are double-spinning, so I'd remove the spin.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -1152,11 +1197,9 @@ public String toString()
-            if (idleTimeoutNanos <= 0)
-                return _jobs.take();
-            return _jobs.poll(idleTimeoutNanos, TimeUnit.NANOSECONDS);
+            return (idleTimeoutNanos <= 0) ? _jobs.take() : _jobs.poll(idleTimeoutNanos, TimeUnit.NANOSECONDS);
```
They were on different lines to help debugging/stacktraces/flamegraphs.

sbordet reviewed (changes requested):
I don't mind the additional implementation in BAQ, but I don't like the solution.
I think the JDK solution is better and simpler.
Can we not use that solution instead?

gregw committed:
Wait for starting thread for bound queue in QTP.

Trialing LinkedTransferQueue as well

gregw committed:
Wait for starting thread for bound queue in QTP.

Trialing LinkedTransferQueue as well - bad idea. Order of magnitude slower in jmh tests

gregw commented:
> I think the JDK solution is better and simpler.
> Can we not use that solution instead?

I tried both ArrayBlockingQueue and LinkedTransferQueue in our QTP jmh tests and they were both an order of magnitude slower (30k/s vs 300k/s).
I also tried a variety of AI suggested queues, including third party ones and got a bit closer (200k/s vs 300k/s) but BAQ is still fastest for QTP

gregw committed:
updates from review

gregw committed:
Embrace the signalAll approach

gregw committed:
Embrace the signalAll approach

gregw committed:
Revert to a cascaded signal one strategy

gregw committed:
fix javadoc

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/13100/qtpRejectedJobs

gregw committed:
updates from review

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -514,25 +629,75 @@ public void add(int index, E e)
+                int tail = _indexes[TAIL_OFFSET];
+                _elements[tail] = e;
+                _indexes[TAIL_OFFSET] = (tail + 1) % _elements.length;
+                wasEmpty = _size.getAndIncrement() == 0;
+                if (size + 1 < _maxCapacity)
```
I think you should "refresh" `size` from the `_size` field when you increment.

gregw committed:
updates from review

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -670,8 +673,9 @@ else if (_growCapacity > 0 && size == _elements.length)
                 _indexes[TAIL_OFFSET] = (tail + 1) % _elements.length;
-                wasEmpty = _size.getAndIncrement() == 0;
-                if (size + 1 < _maxCapacity)
+                int oldSize = _size.getAndIncrement();
+                wasEmpty = oldSize == 0;
+                if (oldSize + 1 < _maxCapacity)
```
This pattern is being repeated, so maybe it should be extracted to a new private method, like `boolean tailLockedIncrementSizeAndSignalIfNotFull()`?

That method could also add a lock held assertion.

sbordet replied:
I prefer the code not to be moved away to another method, so all the code is in the place it belongs to.
In this case, I don't mind a little duplication to get more clarity.

gregw replied:
There is a lot of code duplication.  I had refactored into some common methods, but @sbordet asked me to inline them.
However, I'm not exactly sure inlining does give clarity.  I will named one liner can be clearer than 5 repeated lines that you can't tell if they are exactly the same as the next repetition or not.

I'll have a try to see if I can find a mid ground.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -272,9 +272,9 @@ public E poll(long time, TimeUnit unit) throws InterruptedException
-            int size = _size.getAndDecrement();
-            wasFull = size == _maxCapacity;
-            if (size > 1)
+            int oldSize = _size.getAndDecrement();
+            wasFull = oldSize == _maxCapacity;
+            if (oldSize > 1)
```
This pattern is being repeated, so maybe it should be extracted to a new private method, like `boolean headLockedDecrementSizeAndSignalIfNotEmpty()`?

That method could also add a lock held assertion.

sbordet replied:
Ditto

gregw replied:
ditto

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
In `add()` line 596 of the new version, `size = _size.incrementAndGet();` could be moved down right before the checks for notEmpty and notFull signaling is done so that related changes are co-located; like in `remove()` line 361.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -216,7 +218,7 @@ public E poll()
         E e = null;
         boolean wasFull = false;
-        _headLock.lock(); // Size cannot shrink
+        _headLock.lock(); // Size can grow but not shrink
```
Maybe that comment should be moved next to the `_headLock` field, and add the corollary `Size can shrink but not grow when this lock is held` next to the `_tailLock` field?

sbordet replied:
I'm fine with the comment at the field, but a reminder here helps.

lorban reviewed (changes requested):
I only reviewed the BAQ so far. I think it is now correct, but we could extract all the size-modifying logic to a trio of methods for extra safety and clarity.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -407,34 +464,87 @@ public boolean offer(E e)
-            }
-        }
+        if (wasEmpty)
+            // We signalAll here because there may be many threads waiting to take,
+            // this is more efficient than always grabbing the head lock to always signal 1.
```
Remove obsolete comment.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -407,34 +464,87 @@ public boolean offer(E e)
+        }
+
+        if (wasEmpty)
+            // We signalAll here because there may be many threads waiting to take,
+            // this is more efficient than always grabbing the head lock to always signal 1.
```
Remove obsolete comment.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -770,54 +926,60 @@ public int getMaxCapacity()
-            {
-                _headLock.unlock();
-            }
+    private void signalAll(Lock lock, Condition condition)
```
Never used, remove.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -764,7 +762,10 @@ public int getMaxAvailableThreads()
     public double getUtilizationRate()
     {
-        return (double)getUtilizedThreads() / (getMaxThreads() - getLeasedThreads());
+        int available = getMaxThreads() - getLeasedThreads();
+        if (available <= 0)
+            throw new IllegalStateException("too many leased threads");
+        return (double)getUtilizedThreads() / available;
```
Unnecessary, cannot be that `maxThread==leasedThreads` since `ThreadPoolBudget` avoids that.
I'd restore the old code.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -804,32 +803,78 @@ public void execute(Runnable job)
+            }
+            catch (UnsupportedOperationException uoe)
+            {
+                LOG.warn("Detected bounded thread pool queue {} without put() support", _jobs.getClass());
```
If `put()` is not implemented, we will get this WARN a million times in the logs.

gregw replied:
No, only once because of the boolean below

gregw replied:
No it was protected from that.... but removed now anyway

gregw committed:
updates from review

gregw commented:
@sbordet I sided with @lorban over having small utility methods.   Having duplicate code was OK when there were only 2 methods, but now we are implementing all the methods, I think it is clearer to have those private methods.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -770,57 +897,107 @@ public int getMaxCapacity()
     {
-        if (_growCapacity <= 0)
-            return false;
+        final int head = _indexes[HEAD_OFFSET];
```
Please add an `assert` to check that both locks are held.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -770,57 +897,107 @@ public int getMaxCapacity()
+     */
+    private void lockedAddToTail(E e)
+    {
+        int tail = _indexes[TAIL_OFFSET];
```
Please assert that the lock is held.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -770,57 +897,107 @@ public int getMaxCapacity()
+     */
+    private E lockedTakeFromHead()
+    {
+        final int head = _indexes[HEAD_OFFSET];
```
ditto

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -770,57 +897,107 @@ public int getMaxCapacity()
+     */
+    private boolean lockedIncrementSize()
+    {
+        boolean wasEmpty;
```
ditto

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -770,57 +897,107 @@ public int getMaxCapacity()
+     */
+    private boolean lockedDecrementSize()
+    {
+        boolean wasFull;
```
ditto

lorban commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/thread/QueuedThreadPoolTest.java:
```diff
@@ -1158,6 +1160,57 @@ public void testRealistic() throws Exception
+    @Test
+    public void testBoundedQueue() throws Exception
+    {
+        LoggerFactory.getLogger(QueuedThreadPool.class).warn("Bound queue warnings expected");
```
You could silence the warnings of BAQ instead.

gregw replied:
I like seeing such non-verbose (no stack trace) warnings in the log

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -270,10 +317,10 @@ public E remove()
         try
         {
-            _headLock.lock();
+            _headLock.lock(); // _size can grow but not shrink
```
I'd change the comment to: `both locks are held, _size can either grow or shrink`

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -325,10 +376,10 @@ public E remove(int index)
         try
         {
-            _headLock.lock();
+            _headLock.lock(); // _size can grow but not shrink
```
ditto

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -380,61 +431,101 @@ public boolean offer(E e)
+            if (_growCapacity > 0 && size == _elements.length)
             {
-                _headLock.lock();
+                _headLock.lock(); // _size can grow but not shrink
```
ditto

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -380,61 +431,101 @@ public boolean offer(E e)
+                // Otherwise, can we grow?
+                else if (_growCapacity > 0 && size == _elements.length)
+                {
+                    _headLock.lock(); // _size can grow but not shrink
```
ditto

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -452,13 +543,13 @@ public void add(int index, E e)
         try
         {
-            _headLock.lock();
+            _headLock.lock(); // _size can grow but not shrink
```
ditto

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -514,71 +613,96 @@ public void add(int index, E e)
+                // Otherwise, can we grow?
+                else if (_growCapacity > 0 && size == _elements.length)
+                {
+                    _headLock.lock(); // _size can grow but not shrink
```
ditto

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -591,10 +715,10 @@ public int drainTo(Collection<? super E> c)
         try
         {
-            _headLock.lock();
+            _headLock.lock(); // _size can grow but not shrink
```
ditto

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -647,10 +774,10 @@ public int drainTo(Collection<? super E> c, int maxElements)
         try
         {
-            _headLock.lock();
+            _headLock.lock(); // _size can grow but not shrink
```
ditto

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -678,10 +805,10 @@ public E set(int index, E e)
         try
         {
-            _headLock.lock();
+            _headLock.lock(); // _size can grow but not shrink
```
ditto

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -709,10 +836,10 @@ public E set(int index, E e)
         try
         {
-            _headLock.lock();
+            _headLock.lock(); // _size can grow but not shrink
```
ditto

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java:
```diff
@@ -67,11 +107,23 @@ public class BlockingArrayQueue<E> extends AbstractList<E> implements BlockingQu
+     * Lock for tail operations.
+     * The tail lock is only acquired when adding entries to the queue, thus when held, the size can grow but not shrink.
+     */
+    private final Lock _tailLock = new ReentrantLock();
```
Could you please add a comment to say that when both locks are needed, they must always be taken in order: tail first, head second.

lorban reviewed (changes requested):
Looks good, except for a number of places where I'd add/change the comments and add some assertions.

gregw committed:
Assert locks held

sbordet committed:
Introduced AutoLock.lockInterruptibly(), and AutoLock.WithCondition.awaitNanos(long).
Updated BlockingArrayQueue to use AutoLock.WithCondition, for better readability.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed locking.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet reviewed (approved):
I was unhappy with the comments around the locking of BAQ, as they were inconsistent: sometimes referring to whether the size could shrink/grow, etc.

Replaced the locking using `AutoLock.WithCondition`, much more readable.

LGTM.
